exercises/grep/example.js

```diff
@@ -0,0 +1,65 @@
+var fs = require('fs');
+var readline = require('readline');
+
+var FLAGS = {
+  lineNumbers: '-n',
+  onlyFilenames: '-l',
+  anyCase: '-i',
+  invert: '-v',
+  entireLines: '-x'
+  // fileNames: internal option
+};
+
+var grepFile = function (pattern, options, file, done) {
+  var re = new RegExp(options.entireLines ? '^' + pattern + '$' : pattern, options.anyCase ? 'i' : '');
+  var output = [];
+  var lineNumber = 0;
+  var fstream = fs.createReadStream(file)
+    .on('error', done);
+
+  readline.createInterface({input: fstream})
+    .on('line', function (data) {
+      var line = data;
+      lineNumber++;
+      if (re.test(line) !== options.invert) {
+        if (options.onlyFilenames) {
+          output = [file];
+        } else {
+          if (options.lineNumbers) {
+            line = lineNumber + ':' + line;
+          }
+          if (options.fileNames) {
+            line = file + ':' + line;
+          }
+          output.push(line);
+        }
+      }
+    })
+    .on('close', function () {
+      done(null, output);
+    });
+};
+
+var grep = function (pattern, flags, files, done) {
+  var countdown = files.length;
+  var output = [];
+  var options = {
+    fileNames: files.length > 1
+  };
+  Object.keys(FLAGS).forEach(function (key) {
+    options[key] = flags.indexOf(FLAGS[key]) > -1;
+  });
+
+  files.forEach(function (file, i) {
+    grepFile(pattern, options, file, function (err, match) {
+      if (err) return done(err);
+      output[i] = match;
+      if (--countdown === 0) {
+        done(null, Array.prototype.concat.apply([], output));
```

Comment:
I am a little concerned about this in terms of the API we will provide to users. This demands the solution to accept a `done` function and call it with a specific order of parameters (`err` and `found`) which might not be intuitive to users unless they read the specs thoroughly.

My gut feeling is to keep the API as `file-in --> result-out`. 
What do others think about this @exercism/javascript ?

Reply:
The `file-in --> result-out` pattern would be easier to understand and reason about. Anyway, callbacks are very common on JavaScript, not only in server side Node.js, but in client side broser-JavaScript, so I'm ok with the callback approach.

But, it would be a nice idea to provide a very simple spec to show users the callback interface, I mean, that the implementation should expect a callback with this signature: `function (error, found) {`.

Reply:
And a nice error message if you don't pass proper parameters would be great.